### PR TITLE
fix(workspace): pin corpus clones to the commit, never silently to HEAD

### PR DESCRIPTION
**Pull request by ofri-peretz**
## The bug

`cloneRepo` pins a corpus repository like this:

```ts
git clone --depth 1 --branch <repo.commit> --single-branch …
```

**`--branch` accepts a branch or a tag and never a SHA.** So every SHA-pinned repository failed that clone, fell through to a `--depth 50` clone of the default branch, and was pinned only if the commit happened to be within the last fifty. When it wasn't, the clone kept **HEAD** and printed a warning to stdout.

## How it surfaced

The drift check from [#523](https://github.com/ofri-peretz/eslint/pull/523) on the first CI run of ILB-Corpus-Truth:

```
⚠️  6/107 roots are not at their pinned commit:
     activepieces:              f6cf7003 ≠ 9e58d078
     eslint-config-ts-prefixer: 1edc1606 ≠ b8be3fa6
     ghostfolio:                026464ad ≠ 5bbdf52e
     immich:                    19972326 ≠ cafd6c7c
     novu:                      4926ef59 ≠ d28f579d
     …and 1 more
```

That is the source of the discrepancy #523 was built to detect but could not explain: **107,384 files in one run, 119,415 in another**, over a corpus that reported itself as pinned. Every per-rule comparison drawn across those runs was measuring different code — and the committed baseline was recorded from one of them.

## The fix

- **`fetch --depth 1 origin <sha>`** into a fresh `git init`, then `checkout FETCH_HEAD`. GitHub serves an arbitrary SHA this way, so it works however far behind the pin is.
- **A cached clone that cannot be re-pinned is deleted and re-cloned**, not reused — a cache quietly holding a different commit is the same defect one layer down.
- **No HEAD fallback.** It throws. A benchmark that measures a different commit than the one it reports is worse than one that stops: the numbers look fine and every conclusion drawn from them is wrong.
- Tag and branch pins still clone the ordinary way, so `ILB-Wild`'s `v15.1.0`-style entries are unaffected.

## Test plan

- [x] **Verified against the exact repositories that failed in CI.** All four now check out at their pinned SHA:
  ```
  ✅ activepieces               9e58d078 == pin
  ✅ eslint-config-ts-prefixer  b8be3fa6 == pin
  ✅ ghostfolio                 5bbdf52e == pin
  ✅ novu                       d28f579d == pin
  ```
- [x] `benchmarks` suite green — 160 tests.
- [x] Typecheck clean.

No unit test is added for `cloneRepo` itself: every meaningful assertion needs the network, and a flaky network test in CI is worse than none. The real regression guard is the drift check in ILB-Corpus-Truth, which now fails loudly and names each unpinned root — it is what caught this, and it will catch the next one.

## Why it matters beyond one bench

`cloneRepo` is shared by every bench that pins a corpus, **ILB-Wild included**. Any of them could have been silently measuring a moving target. After this, a corpus is either at its pins or the run stops.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Repository-based benchmarks now use the exact requested commit.
  * Invalid or incomplete cached repositories are discarded and refreshed automatically.
  * Prevented fallback to an incorrect revision when the specified commit cannot be retrieved.

### CODE REVIEW

File: benchmarks/lib/clone-repo.ts [hunk 1]
```diff
@@ -36,32 +65,40 @@ export function cloneRepo(repo: RepoSpec, benchDir: string): string {
...
-        safeGit(['-C', dir, 'checkout', 'FETCH_HEAD'], { allowFail: true });
-        console.log(`   📥 Fetched & checked out ${repo.commit}`);
-      } else {
-        console.log(`   ⚠️  Could not resolve ${repo.commit} on cached clone — using HEAD ${head.slice(0, 7)}`);
-      }
+    const fetched = safeGit(['-C', dir, 'fetch', '--depth', '1', 'origin', repo.commit], { allowFail: true });
+    if (safeGit(['-C', dir, 'rev-parse', 'FETCH_HEAD'], { allowFail: true }).trim()) {
+      safeGit(['-C', dir, 'checkout', '--quiet', 'FETCH_HEAD'], { allowFail: true });
+      console.log(`   📥 Re-pinned cached clone to ${repo.commit.slice(0, 7)}`);
       return dir;
```
[the remark this answers is not in the document]
Comment (ofri-peretz) [reply benchmarks/lib/clone-repo.ts]: Correct, and it is this function's own defect one layer down — thank you.

`FETCH_HEAD` survives an earlier fetch, so checking that it merely *exists* after a failed fetch reads a stale ref, checks out whatever was fetched last, and returns it as if it matched the pin. "Looks pinned, isn't" is precisely what this PR exists to remove, and I reproduced it in the cached path too.

Fixed by verifying the **result** rather than the attempt: `isPinnedTo` answers from `rev-parse HEAD`, which is the only trustworthy evidence — a fetch can fail while leaving a usable `FETCH_HEAD`, and a `checkout` can fail silently through `allowFail`. Tag and branch pins resolve through `rev-parse <ref>` so they compare like for like. Every path now returns only after that check.

Verified all three against a real repository:

```
⬇️  Cloning ghostfolio@5bbdf52…     fresh:  PINNED
📂 Cached at 5bbdf52                cached: PINNED
🧹 Cached clone cannot reach 0000000 — re-cloning
                                    bogus:  threw ✅
```
